haskell/cabal.bzl

```diff
@@ -243,3 +243,225 @@ However, using a plain `haskell_library` sometimes leads to better
 build times, and does not require drafting a `.cabal` file.
 
 """
+
+def _chop_version(name):
+    """Remove any version component from the given package name."""
+    return name.rpartition("-")[0]
+
+# Temporary hardcoded list of core libraries. This will no longer be
+# necessary once Stack 2.0 is released.
+#
+# TODO remove this list and replace it with Stack's --global-hints
+# mechanism.
+_CORE_PACKAGES = [
+    "Cabal",
+    "array",
+    "base",
+    "binary",
+    "bytestring",
+    "containers",
+    "deepseq",
+    "directory",
+    "filepath",
+    "ghc",
+    "ghc-boot",
+    "ghc-boot-th",
+    "ghc-compact",
+    "ghc-heap",
+    "ghc-prim",
+    "ghci",
+    "haskeline",
+    "hpc",
+    "integer-gmp",
+    "libiserv",
+    "mtl",
+    "parsec",
+    "pretty",
+    "process",
+    "rts",
+    "stm",
+    "template-haskell",
+    "terminfo",
+    "text",
+    "time",
+    "transformers",
+    "unix",
+    "xhtml",
+]
+
+def _compute_dependency_graph(repository_ctx, versioned_packages, unversioned_packages):
+    """Given a list of root packages, compute a dependency graph.
+
+    Returns:
+      dependencies: adjacency list of packages, represented as a dictionary.
+      transitive_unpacked_sdists: directory names of unpacked source distributions.
+
+    """
+
+    if not versioned_packages and not unversioned_packages:
+        return ({}, [])
+
+    # Unpack all given packages, then compute the transitive closure
+    # and unpack anything in the transitive closure as well.
+    stack_cmd = repository_ctx.which("stack")
+    if not stack_cmd:
+        fail("Cannot find stack command in your PATH.")
+    stack = [stack_cmd, "--no-nix", "--skip-ghc-check", "--system-ghc"]
+    if versioned_packages:
+        _execute_or_fail_loudly(repository_ctx, stack + ["unpack"] + versioned_packages)
+    stack = [stack_cmd, "--no-nix", "--skip-ghc-check", "--system-ghc", "--resolver", repository_ctx.attr.snapshot]
```

Comment:
As this is a repository rule this just picks whatever `ghc` is in `$PATH`, correct? Does this have any impact on the generated dependency graph? E.g. conditional dependencies based on the `ghc` version, or similar? If so then this seems like another iteration of https://github.com/FormationAI/hazel/issues/80.

Reply:
The nice thing about using Stackage is that in principle they deal with this for us, provided we stick to using the same GHC version as Stackage does for any given snapshot.